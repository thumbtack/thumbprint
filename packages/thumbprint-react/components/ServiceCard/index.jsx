import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Image from '../Image/index.jsx';
import styles from './index.module.scss';

const ServiceCardImage = forwardRef((props, outerRef) => {
    const { url, src, sources, alt } = props;
    return (
        <Image
            className={styles.image}
            sources={sources}
            containerAspectRatio={8 / 5}
            src={url || src} // `url` deprecated
            alt={alt}
            ref={outerRef}
        />
    );
});

function ServiceCardTitle({ children }) {
    return (
        <div className={styles.title} title={children}>
            {children}
        </div>
    );
}

function ServiceCardDescription({ iconColor, icon, children }) {
    return (
        <div
            className={classNames(styles.descriptionWrap, {
                [styles.iconBlue]: iconColor === 'blue',
                [styles.iconGreen]: iconColor === 'green',
            })}
        >
            {icon}
            <div className={styles.description}>{children}</div>
        </div>
    );
}

export default function ServiceCard({ url, children, onClick, shouldOpenInNewTab }) {
    return (
        <a
            href={url}
            onClick={onClick}
            target={shouldOpenInNewTab ? '_blank' : '_self'}
            className={styles.root}
        >
            {children}
        </a>
    );
}

ServiceCard.propTypes = {
    /**
     * URL pointing to the card link destination.
     */
    url: PropTypes.string.isRequired,
    /**
     * Accepts content of `ServiceCardImage`, `ServiceCardTitle`, `ServiceCardDescription`.
     */
    children: PropTypes.node.isRequired,
    /**
     * Handler for click events
     */
    onClick: PropTypes.func,
    /**
     * Opens the URL in a new tab when clicked.
     */
    shouldOpenInNewTab: PropTypes.bool,
};

ServiceCard.defaultProps = {
    onClick: null,
    shouldOpenInNewTab: false,
};

ServiceCardImage.propTypes = {
    /**
     * URL pointing to image to be displayed. This image must have an aspect ratio of 8:5.
     * @deprecated Use `src` instead of `url`.
     */
    url: PropTypes.string,
    /**
     * URL pointing to image to be displayed. This image must have an aspect ratio of 8:5.
     */
    src: PropTypes.string,
    /**
     * Allows the browser to choose the best file format and image size based on the device screen
     * density and the width of the rendered image. Images must have an aspect ratio of 8:5.
     */
    sources: PropTypes.arrayOf(
        PropTypes.shape({
            type: PropTypes.oneOf(['image/webp', 'image/jpeg', 'image/png', 'image/gif'])
                .isRequired,
            srcSet: PropTypes.string.isRequired,
        }),
    ),
    /**
     * Image alt tag that's passed to `aria-label` for better accessibility.
     */
    alt: PropTypes.string,
};

ServiceCardImage.defaultProps = {
    alt: undefined,
    sources: undefined,
    url: undefined,
    src: undefined,
};

// Needed because of the `forwardRef`.
ServiceCardImage.displayName = 'ServiceCardImage';

ServiceCardTitle.propTypes = {
    /**
     * Service title.
     */
    children: PropTypes.node.isRequired,
};

ServiceCardDescription.propTypes = {
    /**
     * Service description or other service info.
     */
    children: PropTypes.node.isRequired,
    /**
     * Icon from [Thumbprint Icons](/icons/) to render within the
     * service card. It must be one of the small icons.
     */
    icon: PropTypes.node,
    /**
     * Color options for icon. Only required if you want to override default color.
     */
    iconColor: PropTypes.oneOf(['blue', 'green']),
};

ServiceCardDescription.defaultProps = {
    icon: undefined,
    iconColor: 'blue',
};

export { ServiceCardDescription, ServiceCardTitle, ServiceCardImage };
