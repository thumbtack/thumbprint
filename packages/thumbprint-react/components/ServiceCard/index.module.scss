@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

.root {
    color: inherit;
    display: block;

    &:hover {
        color: inherit;
    }
}

.image {
    border-radius: $tp-border-radius__base;
    margin-bottom: $tp-space__2;
}

.title {
    font-weight: 700;
    font-size: $tp-font__body__1__size;
    line-height: $tp-font__body__1__line-height;
    // Uses line-height to calculate max-height.
    max-height: $tp-font__body__1__line-height * 2;
    margin-bottom: $tp-space__1;
    overflow: hidden;
    // For webkit browsers that support multiline truncation.
    /* autoprefixer: off */
    display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
}

.descriptionWrap {
    display: flex;
    align-items: flex-start;
}

.description {
    font-size: $tp-font__body__3__size;
    line-height: $tp-font__body__3__line-height;
    color: $tp-color__black-300;
}

.descriptionWrap svg {
    flex: none;
    margin-right: $tp-space__1;
}

.iconBlue svg {
    color: $tp-color__blue;
}

.iconGreen svg {
    color: $tp-color__green;
}
