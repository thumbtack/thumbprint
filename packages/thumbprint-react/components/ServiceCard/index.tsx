import React, { forwardRef } from 'react';
import classNames from 'classnames';
import Image from '../Image/index';
import styles from './index.module.scss';

type ImageSource = {
    type: 'image/webp' | 'image/jpeg' | 'image/png' | 'image/gif';
    srcSet: string;
};

interface ServiceCardImagePropTypes {
    /**
     * URL pointing to image to be displayed. This image must have an aspect ratio of 8:5.
     * @deprecated Use `src` instead of `url`.
     */
    url?: string;
    /**
     * URL pointing to image to be displayed. This image must have an aspect ratio of 8:5.
     */
    src: string;
    /**
     * Allows the browser to choose the best file format and image size based on the device screen
     * density and the width of the rendered image. Images must have an aspect ratio of 8:5.
     */
    sources?: ImageSource[];
    /**
     * Image alt tag that's passed to `aria-label` for better accessibility.
     */
    alt?: string;
}

const ServiceCardImage = forwardRef<HTMLElement, ServiceCardImagePropTypes>(
    (props: ServiceCardImagePropTypes, outerRef) => {
        const { url, src, sources, alt } = props;
        return (
            <Image
                className={styles.image}
                sources={sources}
                containerAspectRatio={8 / 5}
                src={url || src} // `url` deprecated
                alt={alt}
                ref={outerRef}
            />
        );
    },
);

// Needed because of the `forwardRef`.
ServiceCardImage.displayName = 'ServiceCardImage';

interface ServiceCardTitlePropTypes {
    children: string;
}

function ServiceCardTitle({ children }: ServiceCardTitlePropTypes): JSX.Element {
    return (
        <div className={styles.title} title={children}>
            {children}
        </div>
    );
}

interface ServiceCardDescriptionPropTypes {
    /**
     * Service description or other service info.
     */
    children: React.ReactNode;
    /**
     * Icon from [Thumbprint Icons](/icons/) to render within the
     * service card. It must be one of the small icons.
     */
    icon?: React.ReactNode;
    /**
     * Color options for icon. Only required if you want to override default color.
     */
    iconColor?: 'blue' | 'green';
}

function ServiceCardDescription({
    iconColor = 'blue',
    icon,
    children,
}: ServiceCardDescriptionPropTypes): JSX.Element {
    return (
        <div
            className={classNames(styles.descriptionWrap, {
                [styles.iconBlue]: iconColor === 'blue',
                [styles.iconGreen]: iconColor === 'green',
            })}
        >
            {icon}
            <div className={styles.description}>{children}</div>
        </div>
    );
}

interface ServiceCardPropTypes {
    /**
     * URL pointing to the card link destination.
     */
    url: string;
    /**
     * Accepts content of `ServiceCardImage`, `ServiceCardTitle`, `ServiceCardDescription`.
     */
    children: React.ReactNode;
    /**
     * Handler for click events
     */
    onClick?: () => void;
    /**
     * Opens the URL in a new tab when clicked.
     */
    shouldOpenInNewTab?: boolean;
}

export default function ServiceCard({
    url,
    children,
    onClick,
    shouldOpenInNewTab = false,
}: ServiceCardPropTypes): JSX.Element {
    return (
        <a
            href={url}
            onClick={onClick}
            target={shouldOpenInNewTab ? '_blank' : '_self'}
            className={styles.root}
        >
            {children}
        </a>
    );
}

export { ServiceCardDescription, ServiceCardTitle, ServiceCardImage };
