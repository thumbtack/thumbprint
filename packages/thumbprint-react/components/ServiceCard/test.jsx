import React from 'react';
import { mount } from 'enzyme';
import ServiceCard, { ServiceCardImage, ServiceCardTitle, ServiceCardDescription } from './index';

describe('ServiceCard', () => {
    test('renders an href when supplied', () => {
        const wrapper = mount(
            <ServiceCard url="https://www.thumbtack.com/k/massage/near-me/">Lorem</ServiceCard>,
        );
        expect(wrapper.find('a').prop('href')).toBe('https://www.thumbtack.com/k/massage/near-me/');
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ServiceCardImage', () => {
    test('renders an aria-label when supplied by alt', () => {
        const wrapper = mount(<ServiceCardImage alt="Lorem" url="" />);
        expect(wrapper.find('.image').prop('aria-label')).toBe('Lorem');
        expect(wrapper).toMatchSnapshot();
    });

    test('adds lazyload class for lazysizes to work', () => {
        const wrapper = mount(<ServiceCardImage alt="Lorem" url="" />);
        expect(wrapper.find('.lazyload')).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('renders an background image when supplied', () => {
        const wrapper = mount(
            <ServiceCardImage alt="" url="https://www.thumbtack.com/image.png" />,
        );
        expect(wrapper.find('.image').prop('data-bg')).toBe('https://www.thumbtack.com/image.png');
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ServiceCardTitle', () => {
    test('renders a title when supplied', () => {
        const wrapper = mount(<ServiceCardTitle>Lorem</ServiceCardTitle>);
        expect(wrapper.text()).toBe('Lorem');
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ServiceCardDescription', () => {
    test('renders a description when supplied', () => {
        const wrapper = mount(<ServiceCardDescription>Lorem</ServiceCardDescription>);
        expect(wrapper.text()).toBe('Lorem');
        expect(wrapper).toMatchSnapshot();
    });

    test('renders a icon and color is blue', () => {
        const wrapper = mount(
            <ServiceCardDescription icon={<svg />}>Lorem</ServiceCardDescription>,
        );
        expect(wrapper.find('.descriptionWrap svg'));
        expect(wrapper.find('.descriptionWrap').hasClass('iconBlue'));
        expect(wrapper).toMatchSnapshot();
    });

    test('renders icon color', () => {
        const wrapper = mount(
            <ServiceCardDescription icon="" iconColor="green">
                Lorem
            </ServiceCardDescription>,
        );
        expect(wrapper.find('.descriptionWrap').hasClass('iconGreen'));
        expect(wrapper).toMatchSnapshot();
    });
});
