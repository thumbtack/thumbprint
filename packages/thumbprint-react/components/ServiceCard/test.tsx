import React from 'react';
import { mount } from 'enzyme';
import ServiceCard, { ServiceCardImage, ServiceCardTitle, ServiceCardDescription } from './index';

// TODO(giles): All this `react-intersection-observer` mocking stuff is used in a few unit test
// suites (anything that uses our lazy-loading `Image`). Move to a shared utility so we don't have
// all this noise at the top of every file.
type UseInViewReturnType = [() => void, true];
type InViewChildrenMockType = (arg1: { inView: true; ref: null }) => void;
type ReactIntersectionObserverMockType = {
    InView: ({ children }: { children: InViewChildrenMockType }) => void;
    useInView: () => UseInViewReturnType;
};

jest.mock(
    'react-intersection-observer',
    (): ReactIntersectionObserverMockType => ({
        InView: ({ children }: { children: InViewChildrenMockType }): void =>
            children({ inView: true, ref: null }),
        useInView: (): UseInViewReturnType => [(): void => {}, true],
    }),
);

beforeEach((): void => {
    window.IntersectionObserver = (true as unknown) as (typeof window.IntersectionObserver);
});

describe('ServiceCard', () => {
    test('renders an href when supplied', () => {
        const wrapper = mount(
            <ServiceCard url="https://www.thumbtack.com/k/massage/near-me/">Lorem</ServiceCard>,
        );
        expect(wrapper.find('a').prop('href')).toBe('https://www.thumbtack.com/k/massage/near-me/');
        expect(wrapper).toMatchSnapshot();
    });
    test('renders an href with the proper target', () => {
        const wrapper = mount(
            <ServiceCard url="https://www.thumbtack.com/k/massage/near-me/" shouldOpenInNewTab>
                Lorem
            </ServiceCard>,
        );
        expect(wrapper.find('a').prop('target')).toBe('_blank');
    });
    test('renders an href with a click handler', () => {
        const onClick = jest.fn();
        const wrapper = mount(
            <ServiceCard url="https://www.thumbtack.com/k/massage/near-me/" onClick={onClick}>
                Lorem
            </ServiceCard>,
        );
        wrapper.find('a').simulate('click');
        expect(onClick).toHaveBeenCalled();
    });
});

describe('ServiceCardImage', () => {
    test('render works', () => {
        const wrapper = mount(
            <ServiceCardImage alt="duck duck goose" src="https://www.thumbtack.com/image.png" />,
        );

        expect(wrapper).toMatchSnapshot();
    });
});

describe('ServiceCardTitle', () => {
    test('renders a title when supplied', () => {
        const wrapper = mount(<ServiceCardTitle>Lorem</ServiceCardTitle>);
        expect(wrapper.text()).toBe('Lorem');
        expect(wrapper).toMatchSnapshot();
    });
});

describe('ServiceCardDescription', () => {
    test('renders a description when supplied', () => {
        const wrapper = mount(<ServiceCardDescription>Lorem</ServiceCardDescription>);
        expect(wrapper.text()).toBe('Lorem');
        expect(wrapper).toMatchSnapshot();
    });

    test('renders a icon and color is blue', () => {
        const wrapper = mount(
            <ServiceCardDescription icon={<svg />}>Lorem</ServiceCardDescription>,
        );
        expect(wrapper.find('.descriptionWrap svg'));
        expect(wrapper.find('.descriptionWrap').hasClass('iconBlue'));
        expect(wrapper).toMatchSnapshot();
    });

    test('renders icon color', () => {
        const wrapper = mount(
            <ServiceCardDescription icon="" iconColor="green">
                Lorem
            </ServiceCardDescription>,
        );
        expect(wrapper.find('.descriptionWrap').hasClass('iconGreen'));
        expect(wrapper).toMatchSnapshot();
    });
});
