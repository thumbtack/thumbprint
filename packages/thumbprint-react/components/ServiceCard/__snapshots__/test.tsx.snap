// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceCard renders an href when supplied 1`] = `
<ServiceCard
  url="https://www.thumbtack.com/k/massage/near-me/"
>
  <a
    className="root"
    href="https://www.thumbtack.com/k/massage/near-me/"
    target="_self"
  >
    Lorem
  </a>
</ServiceCard>
`;

exports[`ServiceCardDescription renders a description when supplied 1`] = `
<ServiceCardDescription>
  <div
    className="descriptionWrap iconBlue"
  >
    <div
      className="description"
    >
      Lorem
    </div>
  </div>
</ServiceCardDescription>
`;

exports[`ServiceCardDescription renders a icon and color is blue 1`] = `
<ServiceCardDescription
  icon={<svg />}
>
  <div
    className="descriptionWrap iconBlue"
  >
    <svg />
    <div
      className="description"
    >
      Lorem
    </div>
  </div>
</ServiceCardDescription>
`;

exports[`ServiceCardDescription renders icon color 1`] = `
<ServiceCardDescription
  icon=""
  iconColor="green"
>
  <div
    className="descriptionWrap iconGreen"
  >
    <div
      className="description"
    >
      Lorem
    </div>
  </div>
</ServiceCardDescription>
`;

exports[`ServiceCardImage render works 1`] = `
<ServiceCardImage
  alt="duck duck goose"
  src="https://www.thumbtack.com/image.png"
>
  <Image
    alt="duck duck goose"
    className="image"
    containerAspectRatio={1.6}
    src="https://www.thumbtack.com/image.png"
  >
    <picture
      className="picture image"
    >
      <img
        alt="duck duck goose"
        className="imageStart"
        onError={[Function]}
        onLoad={[Function]}
        sizes="0px"
        src="https://www.thumbtack.com/image.png"
        style={
          Object {
            "height": 0,
            "overflow": "hidden",
            "paddingTop": "62.5%",
          }
        }
      />
    </picture>
    <noscript>
      <img
        alt="duck duck goose"
        src="https://www.thumbtack.com/image.png"
      />
    </noscript>
  </Image>
</ServiceCardImage>
`;

exports[`ServiceCardTitle renders a title when supplied 1`] = `
<ServiceCardTitle>
  <div
    className="title"
    title="Lorem"
  >
    Lorem
  </div>
</ServiceCardTitle>
`;
